Issue2544 ZoneGroup

bravache opened the pull request:
This branch duplicates the changes that were proposed in [https://github.com/lbl-srg/modelica-buildings/tree/issue2544_zonegroup](`issue2544_zonegroup`) without dragging the templates base branch changes with it.

---

This PR is addressing issue #2544 .

## Constraints
The constraints when addressing this issue was to not rely on `for` loops when writing equation or connecting blocks together. 

Another constraints is that the zone groups are not equally sized, and so it is not possible to create a singular class that could do all the routing between zones and groups since it would require to have the output connectors be non-rectangular 2-dimensional arrays.

## Proposed solution
The solution I propose in this PR is to duplicate the signal of the zones by the number of groups, so that each iteration can be connected to a separate group, and then have that group filter only the zones that are present in it. This is a bit clunky, and there might be a better solution, but from separate test, this seems to do the job.

I've introduced one new block for routing: `ZoneStatusDuplicator` which is used to duplicate the zone status signals so that they can be connected to all group status simultaneously. The zones are being filtered in `GroupStatus`.

JayHuLBL commented on Buildings/Controls/OBC/ASHRAE/G36_PR1/Generic/SetPoints/GroupStatus.mo:
```diff
@@ -151,51 +156,119 @@ block GroupStatus "Block that outputs the zone group status"
     annotation (Placement(transformation(extent={{100,-320},{140,-280}}),
       iconTransformation(extent={{100,-210},{140,-170}})));
 
+  CDL.Routing.BooleanArrayFilter zonOccFil(
```
It would a good practice to use the full path, `Buildings.Controls.OBC.CDL....`. Also, the parameters can be finalized:
```
final nin=numZon,
final nout=numZonGro,
final msk=zonGroMsk
```

JayHuLBL replied:
may also put this section to be protected.

JayHuLBL commented on Buildings/Controls/OBC/ASHRAE/G36_PR1/Generic/SetPoints/GroupStatus.mo:
```diff
@@ -151,51 +156,119 @@ block GroupStatus "Block that outputs the zone group status"
     annotation (Placement(transformation(extent={{40,130},{60,150}})));
-  Buildings.Controls.OBC.CDL.Logical.MultiOr mulOr(
-    final nu=numZon)
+  Buildings.Controls.OBC.CDL.Logical.MultiOr mulOr(nu=numZonGro)
```
`final nu=numZonGro`

JayHuLBL commented on Buildings/Controls/OBC/ASHRAE/G36_PR1/Generic/SetPoints/ZoneStatusDuplicator.mo:
```diff
@@ -0,0 +1,364 @@
+  parameter Integer nGro(final min=1)=1 "Number of groups in output";
+
+
+  CDL.Interfaces.BooleanInput                        zonOcc[nZon]
```
full path

JayHuLBL commented on Buildings/Controls/OBC/ASHRAE/G36_PR1/Generic/SetPoints/ZoneStatusDuplicator.mo:
```diff
@@ -0,0 +1,364 @@
+    annotation (Line(points={{12,-440},{60,-440}}, color={0,0,127}));
+  connect(uWinDup.y, yWin)
+    annotation (Line(points={{12,-520},{60,-520}}, color={255,0,255}));
+  annotation (Icon(coordinateSystem(preserveAspectRatio=false, extent={{-40,-180},
```
missing `defaultComponentName`.

JayHuLBL commented on Buildings/Controls/OBC/CDL/Routing/RealArrayFilter.mo:
```diff
@@ -0,0 +1,64 @@
+  parameter Integer nout "Size of output array";
+  parameter Boolean msk[nin]=fill(true,nin) "Array mask";
+
+  Interfaces.RealInput    u[nin]
```
avoid the empty spaces

JayHuLBL commented on Buildings/Controls/OBC/CDL/Routing/Validation/IntegerArrayReplicator.mo:
```diff
@@ -0,0 +1,58 @@
+within Buildings.Controls.OBC.CDL.Routing.Validation;
+model IntegerArrayReplicator
+  "Validation model for the IntegerArrayReplicator block"
+  Buildings.Controls.OBC.CDL.Routing.IntegerArrayReplicator
```
improve the line change

JayHuLBL commented on Buildings/Controls/OBC/CDL/Routing/Validation/RealArrayReplicator.mo:
```diff
@@ -0,0 +1,53 @@
+within Buildings.Controls.OBC.CDL.Routing.Validation;
+model RealArrayReplicator
+  "Validation model for the RealArrayReplicator block"
+  Buildings.Controls.OBC.CDL.Routing.RealArrayReplicator
```
line change

JayHuLBL commented on Buildings/Controls/OBC/ASHRAE/G36_PR1/Generic/SetPoints/Validation/ZoneStatusDuplicator.mo:
```diff
@@ -0,0 +1,77 @@
+    zonStaDup(nZon=5, nGro=2)
+              "Zone status duplicator"
+    annotation (Placement(transformation(extent={{0,-40},{20,40}})));
+  CDL.Continuous.Sources.Constant reaInp[6,5](k=fill({1,2,3,4,5}, 6))
```
Prefer to use the full path

JayHuLBL reviewed (changes requested):
@bravache Please see the inline comments. Also, the CI test was not passed. More specifically, the new model `Buildings.Controls.OBC.CDL.Routing.Validation.BooleanArrayFilter` failed to simulate with `Optimica` and `Jmodelica`.

bravache reviewed (commented):
Thanks for the feedback @JayHuLBL . I've addressed your comment, and fixed the issue with the filters block validation not running on Optimica and Jmodelica.
This is ready for your re-review.

JayHuLBL reviewed (changes requested):
@bravache The changes look good. However, we should also update the release note in `Buildings.package.mo` to document the new models and changes.

bravache commented:
@JayHuLBL Added the top-level documentation.

JayHuLBL reviewed (approved):
@bravache The changes look good.

mwetter reviewed (changes requested):
I made minor changes directly on the branch.

I think however that in the new CDL blocks, we should not call them `*Array*` because they work on a vector. If we call them `*Vector*` then in the future we could add similar blocks that act on matrices and call them `*Matrix*` (if these are needed). Hence I suggest that these are renamed and the documentation is revised accordingly.

Note that "Vector" is also used in the Modelica Standard Library.

For consistency, I think we should also rename `CDL.Routing.{Boolean,Integer,Real}Replicator` to `{Boolean,Integer,Real}ScalarReplicator`. Then, we have these blocks that work on scalar and that work on vector, and they are named consistently.

bravache commented:
@mwetter Thanks for the suggestion Michael! I have renamed the `Array` to `Vector` wherever I could find them, and also renamed all the other `Replicator` to `ScalarReplicator`. Note that there are still classes called `BooleanReplicator`, `IntegerReplicator` and `RealReplicator` in the `Buildings.Utilities.Math` package.

mwetter reviewed (changes requested):
@bravache The models `{Boolean,Real,Integer}Replicator`  have been renamed to `{Boolean,Real,Integer}ScalarReplicator` which is good. But this needs to be entered in the conversion script and in the non-backward compatible section of the revision notes in the top-level `package.mo` file.

Afterwards it is ready to be merged.

bravache commented:
@mwetter I had already written a conversion script (unless I missed something) and just added a note in the version release notes.

mwetter commented:
@bravache : You are right, the conversion script is updated. I missed it in my review.

bravache commented:
@mwetter Ready to merge
